internal/pkg/gateway/commit/notifier.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2021 IBM All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package commit
+
+import (
+	"context"
+	"sync"
+
+	"github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/hyperledger/fabric/core/ledger"
+)
+
+// NotificationSupplier obtains a commit notification channel for a specific ledger. It provides an abstraction of the use of
+// Peer, Channel and Ledger to obtain this result, and allows mocking in unit tests.
+type NotificationSupplier interface {
+	CommitNotifications(done <-chan struct{}, channelName string) (<-chan *ledger.CommitNotification, error)
+}
+
+// Notifier provides notification of transaction commits.
+type Notifier struct {
+	supplier         NotificationSupplier
+	lock             sync.Mutex
+	channelNotifiers map[string]*channelLevelNotifier
+	cancel           chan struct{}
+	once             sync.Once
+}
+
+// Notification of a specific transaction commit.
+type Notification struct {
+	BlockNumber    uint64
+	TransactionID  string
+	ValidationCode peer.TxValidationCode
+}
+
+// NewNotifier constructor.
+func NewNotifier(supplier NotificationSupplier) *Notifier {
+	if supplier == nil {
+		return nil
+	}
```

Comment:
This is just getting silly.

Reply:
For my own education, can you explain why this approach suggested by Effective Go and used in standard Go package implementations is wrong in this case? What would be the correct answer here?